pkg/servercounter/lease_counter_test.go

```diff
@@ -0,0 +1,249 @@
+package servercounter
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	coordinationv1 "k8s.io/api/coordination/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	coordinationv1listers "k8s.io/client-go/listers/coordination/v1"
+)
+
+type leaseTemplate struct {
+	durationSecs     int32
+	timeSinceAcquire time.Duration
+	timeSinceRenew   time.Duration
+	labels           map[string]string
+}
+
+func newLeaseFromTemplate(template leaseTemplate) *coordinationv1.Lease {
+	lease := &coordinationv1.Lease{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: template.labels,
+		},
+		Spec: coordinationv1.LeaseSpec{},
+	}
+
+	if template.durationSecs != 0 {
+		lease.Spec.LeaseDurationSeconds = &template.durationSecs
+	}
+	if template.timeSinceAcquire != time.Duration(0) {
+		acquireTime := metav1.NewMicroTime(time.Now().Add(-template.timeSinceAcquire))
```

Comment:
Reading how this is used in the tests, I don't think there is an issue using time.Now().  However if we were doing anything more complicated it would be better to:

1. Have a `var timeNow = time.Now` declaration in the package.  This variable is used in place of directly calling `time.Now()`.
2. Implement a `type controlledTime struct {}` in the test file that allows for manual manipulation of the flow of time.
3. In tests, override `timeNow` to be a `controlledTime`.

This would allow for determinism in unit tests.  Using time.Now() in unit tests is a common gotcha that causes flakeyness

Reply:
Ended up going with #1.

Reply:
sorry I could have explained better, these are all steps for a way to get around using time.Now(), not separate options.

Basically the issue with using time.Now() is that we don't control the test execution timing.  The key interaction is in lease_counter.go `return lastRenewTime.Add(duration).After(time.Now())`.  In the unit tests this is a possible (however unlikely) source of nondeterminism. 

Using the "freshly renewed lease is valid" case as an example with the current code:

1. timeNow is set to some fixed time.
2. newLeaseFromTemplate() is called, which creates a `Lease{ Spec { LeaseDurationSeconds: 1000, AcquireTime: timeNow - 1000s, RenewTime: timeNow - 1s }}`
3. (however unlikely) process running the test gets preempted for 1001 seconds.
4. isLeaseValid() is called.  `lastRenewTime` is set to `timeNow - 1s`.  `duration` is set to 1000. 
 `lastRenewTime.Add(duration).After(time.Now())` returns false.  The test fails.

We can control this nondeterminism by not directly using time.Now() in both lease_counter.go and lease_counter_test.go.  In lease_counter.go, we declare a `var timeNow` function that is initially set to `time.Now` (the function, not a value from calling the function).  We use the `timeNow` function alias to get the current time.  In non-test code, this is equivalent to directly using `time.Now` to get the current time.

In lease_counter_test.go, we add a `type controlledTime struct { t time.Time }` and implement methods `Now() time.Time` and probably something like `Advance(d time.Duration)`.  In each test we construct a controlledTime and set the package `timeNow` to be the `Now` function on the controlledTime (note that this means we can not run tests cases concurrently, if we did we would have multiple tests writing to `timeNow` which would cause nondeterminism).  In the test we would manually call `Advance()` to simulate the passage of time.  Looking back at our example, this would look something like:

1. `t := controlledTime{ time.Unix(1000, 0) }`
2. `timeNow = t.Now`
3. newLeaseFromTemplate() is called, which creates a `Lease{ Spec { LeaseDurationSeconds: 1000, AcquireTime: time.Unix(0, 0), RenewTime: time.Unix(999 0) }}`
4. t.Advance(time.Second)
5. (however unlikely) process running the test gets preempted for 1001 seconds.
6. isLeaseValid() is called.  `lastRenewTime` is set to `time.Unix(999, 0)`.  `duration` is set to 1000. 
 `lastRenewTime.Add(duration).After(timeNow() /* returns time.Unix(1001, 0) */)` returns true.  The test succeeds

Reply:
Ohhhh, understood. Doing this now!